refactor: update botstrap to be more beginner-friendly

onerandomusername opened the pull request:
Update botstrap to be much more userfriendly. This silences noisy logs that we don't want to see or care about when downloading new server.

We also provide a success message when the entire system is successful.

I've also implemented some small caching in the botcore client to only make requests when necessary, and to only make one request to get all of the webhooks. While yes, webhooks are free, I left request logging enabled so the user could see what requests were made. I didn't see a good reason to silence those, both for transparency and because they already provide logging. 

We also provide an invite link if the bot isn't in the guild already, but it is configured.

I've also added support which sets the app's flags to enable message content intents and member intent if they aren't already enabled.

closes #3414

jb3 commented on botstrap.py:
```diff
@@ -1,16 +1,23 @@
 load_dotenv()
-log = get_logger("Config Bootstrapper")
+log = get_logger("botstrap")
+# silence noisy httpcore logger
```
https://peps.python.org/pep-0008/#comments
```suggestion
# Silence noisy httpcore logger
```

jb3 commented on botstrap.py:
```diff
@@ -51,10 +76,10 @@ def __getitem__(self, item: str):
             return super().__getitem__(item)
         except KeyError:
-            log.warning(f"Couldn't find key: {item} in dict: {self.name} ")
+            log.fatal(f"Couldn't find key: {item} in dict: {self.name} ")
```
```suggestion
            log.fatal(f"Couldn't find key: {item} in dict: {self.name}")
```

onerandomusername replied:
https://github.com/python-discord/bot/pull/3417/commits/00c597e28255d2c84f69f7bee81433b2c997a9ec

jb3 commented on botstrap.py:
```diff
@@ -65,64 +90,113 @@ class DiscordClient(Client):
         """A boolean that indicates if a channel is of type GUILD_FORUM."""
-        response = self.get(f"/channels/{channel_id_}")
-        return response.json()["type"] == GUILD_FORUM_TYPE
+        return next(filter(lambda c: c["id"] == channel_id_, self.guild_channels)).get("type", None) == GUILD_FORUM_TYPE
```
This isn't very clean and channel `type` is a mandatory attribute from the API, split it up and remove the unnecessary `get()` usage.

I suggest a new helper method to return a channel from `self.guild_channels` and then this method just has to use that and check the type key.

I'd also remove the unnecessary `_` suffixes from the arguments to these helper methods.

onerandomusername replied:
To be fair, given the template now has the forum channel in it, this could be removed entirely. In my testing the code that upgrades python-help doesn't run when a guild is created from the template.

jb3 commented on botstrap.py:
```diff
@@ -147,37 +220,47 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+        log.warning(
+            f"Please invite with the following URL and rerun this script: https://discord.com/oauth2/authorize?client_id={client_id}&guild_id={GUILD_ID}&scope=bot+applications.commands&permissions=8"
+        )
+        sys.exit(42)
```
Why 42?

jb3 reviewed (changes requested):
Few review comments.

Can we also squash down the history a little bit? I don't think we need 21 commits for the changes here. Obviously keep the constants one separate but I think we can squash down a lot of the other ones.

onerandomusername commented on botstrap.py:
```diff
@@ -225,29 +306,40 @@ def create_webhook(self, name: str, channel_id_: int) -> str:
-    config_str += "emojis_trashcan=🗑️"
+
+    # we add some replacements for some of the emojis used by the bot
+    # this is a best-effort and doesn't replace every emote
```
I was working on reworking this segment to instead clone all of the configured emojis. No longer only making a best-effort approach, it is now full effort.

onerandomusername commented on botstrap.py:
```diff
@@ -1,16 +1,23 @@
 
-from bot.constants import Webhooks, _Categories, _Channels, _Roles
-from bot.log import get_logger
+# we want to filter out the patching send typing notifications from bot core when we import to get constants
```
Updated in https://github.com/python-discord/bot/pull/3417/commits/00c597e28255d2c84f69f7bee81433b2c997a9ec

onerandomusername commented:
@jb3 I've compressed the history and addressed your comments in the updated history. Comments should now follow pep8, formatted with ruff, and silenced pyright.

jb3 commented on botstrap.py:
```diff
@@ -147,37 +219,82 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+
+    def get_emoji_contents(self, emoji_id: str | int) -> bytes | None:
+        """Fetches the image data for an emoji by ID."""
+        # emojis are located at https://cdn.discordapp.com/emojis/821611231663751168.png?size=4096
```
I'd use path-params to document this and where that snowflake comes from (even though it's obvious). Also, not sure the size is necessary given we don't then use it for the subsequent request.
```suggestion
        # Emojis are located at https://cdn.discordapp.com/emojis/{emoji_id}.png
```

jb3 commented on botstrap.py:
```diff
@@ -137,8 +210,7 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
-        server_channels = response.json()
+        server_channels = self.guild_channels
 
         for channel in server_channels:
```
```suggestion
        for channel in self.guild_channels:
```

jb3 commented on botstrap.py:
```diff
@@ -63,66 +79,123 @@ def __getitem__(self, item: str):
+        """Check if the bot is a member of the guild."""
+        try:
+            _ = self.guild_info
+        except HTTPStatusError:
```
Maybe we can also check the status here? Avoids us having problems from a transient Discord error that we should surface in a different way.

jb3 reviewed (changes requested):
Sorry for these comments all being in parts, just keep doing scans and picking up new bits.

Can we also standardise on whether you're using f-strings in this file or using `%s` interpolation? There is a mix currently (I appreciate it was f-strings before which is not correct, but if you're adding some f-strings and some %s strings we may as well unify).

jb3 commented on botstrap.py:
```diff
@@ -137,117 +218,252 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+        response = self.post(
+            f"/guilds/{self.guild_id}/emojis",
+            json=payload,
+            headers={"X-Audit-Log-Reason": f"Creating {new_name} emoji as part of PyDis botstrap"},
```
```suggestion
            headers={"X-Audit-Log-Reason": "Creating emoji as part of PyDis botstrap"},
```

onerandomusername commented on botstrap.py:
```diff
@@ -137,117 +218,252 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+if __name__ == "__main__":
+    botstrap = BotStrapper(guild_id=GUILD_ID, env_file=ENV_FILE)
+    with botstrap:
+        botstrap.run()
```
I would prefer to manage the DiscordClient internally on the botstrap class, even if we use contextlib to manage the context manager. 

DiscordClient should not be used by anything other than the BotStrap class, and an instance of it outside of botstrap doesn't need to exist.

jb3 commented on botstrap.py:
```diff
@@ -137,117 +221,259 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+
+    def write_config_env(self, config: dict[str, dict[str, Any]], env_file: Path) -> None:
+        """Write the configuration to the specified env_file."""
+        # in order to support commented sections, we write the following
```
```suggestion
        # In order to support commented sections, we write the following
```

jb3 commented on botstrap.py:
```diff
@@ -51,78 +71,142 @@ def __getitem__(self, item: str):
     def _raise_for_status(response: Response) -> None:
         response.raise_for_status()
 
+    @property
```
Maybe a good case for a switch to a method with https://docs.python.org/3/library/functools.html#functools.cache

onerandomusername commented on botstrap.py:
```diff
@@ -137,117 +221,259 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+        self.client.__exit__(exc_type, exc_value, traceback)
 
-    config_str += "\n#Categories\n"
+    def upgrade_client(self) -> bool:
```
Added the code that consumes it.

shtlrs commented on botstrap.py:
```diff
@@ -83,11 +83,37 @@ def __init__(self, guild_id: int | str):
+        """
+        # Fetch first to modify, not overwrite
+        current_flags = self.app_info.get("flags", 0)
+        new_flags = current_flags | 1 << 15 | 1 << 19
```
Those bit shifts are also "magic numbers", aren't they ? it'd better to assign each one to a variable indicating the flag.

shtlrs commented on botstrap.py:
```diff
@@ -210,6 +220,17 @@ def create_webhook(self, name: str, channel_id_: int) -> str:
         log.info("Application flags upgraded successfully, and necessary intents are now enabled.")
 
+    if not discord_client.check_if_in_guild():
+        client_id = discord_client.app_info["id"]
```
I think it's better to encapsulate this with a property

jb3 commented on botstrap.py:
```diff
@@ -137,117 +193,289 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+        log.info("Botstrap completed successfully. Updated configuration has been written to %s", ENV_FILE)
+    else:
+        log.info("Botstrap completed successfully. No changes were necessary.")
+    sys.exit(changes_made)
```
Why exit with an error if we make changes?

jb3 commented on botstrap.py:
```diff
@@ -137,117 +193,289 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+    if changes_made:
+        log.info("Botstrap completed successfully. Updated configuration has been written to %s", ENV_FILE)
+    else:
+        log.info("Botstrap completed successfully. No changes were necessary.")
```
Space.
```suggestion
    with botstrap:
        changes_made = botstrap.run()

    if changes_made:
        log.info("Botstrap completed successfully. Updated configuration has been written to %s", ENV_FILE)
    else:
        log.info("Botstrap completed successfully. No changes were necessary.")

```

jb3 commented on botstrap.py:
```diff
@@ -137,117 +193,289 @@ def get_all_channels_and_categories(self) -> tuple[dict[str, str], dict[str, str
+    def get_emoji_contents(self, id_: str | int) -> bytes | None:
+        """Fetches the image data for an emoji by ID."""
+        # Emojis are located at https://cdn.discordapp.com/emojis/{emoji_id}.{ext}
+        response = self.get(f"{self.CDN_BASE_URL}/emojis/{id_!s}.webp")
```
Shouldn't fetch webp if you then submit `image/png`, even if Discord is accepting it it's not right.
